src/main/java/org/opensearch/security/privileges/RestLayerPrivilegesEvaluator.java

```diff
@@ -0,0 +1,160 @@
+
+        final TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
+
+        Set<String> mappedRoles = mapRoles(user, caller);
+
+        presponse.resolvedSecurityRoles.addAll(mappedRoles);
+        final SecurityRoles securityRoles = getSecurityRoles(mappedRoles);
+
+        final boolean isDebugEnabled = log.isDebugEnabled();
+        if (isDebugEnabled) {
+            log.debug("Evaluate permissions for {} on {}", user, clusterService.localNode().getName());
+            log.debug("Action: {}", action0);
+            log.debug("Mapped roles: {}", mappedRoles.toString());
+        }
+
+        if (!securityRoles.impliesClusterPermissionPermission(action0) && !impliesLegacyPermission(action0, securityRoles)) {
+            presponse.missingPrivileges.add(action0);
+            presponse.allowed = false;
+            log.info("No permission match for {} [Action [{}]] [RolesChecked {}]. No permissions for {}",  user, action0,
+                    securityRoles.getRoleNames(), presponse.missingPrivileges);
+        } else {
+            if (isDebugEnabled) {
+                log.debug("Allowed because we have extension permissions for {}", action0);
+            }
+            presponse.allowed = true;
+        }
+        return presponse;
+    }
+
+    public Set<String> mapRoles(final User user, final TransportAddress caller) {
+        return this.configModel.mapSecurityRoles(user, caller);
+    }
+
+    /**
+     * Checks if the route is accessible via legacy naming convention.
+     * We check against cluster permissions because the legacy convention for cluster permissions map 1-1 with a transport
+     * action call initiated via REST API handler. Hence we use the same to allow/block request forwarding to extensions.
+     * This ensures backwards-compatibility
+     *
+     *
+     * NOTE: THIS CHECK WILL BE REMOVED ONCE ALL ACTIONS HAVE BEEN MIGRATED TO THE NEW CONVENTION
+     *
+     * E.g For extension `hw`, following are two possible ways actions an be defined in roles:
+     *
+     * extension_hw_full:
+     *   reserved: true
+     *   cluster_permissions:
+     *     - 'extension:hw/greet'
+     *
+     * legacy_hw_full:
+     *   reserved: true
+     *   cluster_permissions:
+     *     - 'cluster:admin/opensearch/hw/greet'
+     *
+     *
+     * @param action - The action to be checked against its legacy version
+     * @return true, if a legacy version was found and validated, false otherwise
+     */
+    private boolean impliesLegacyPermission(String action, SecurityRoles roles) {
+        boolean isAlreadyLegacy = action.startsWith("cluster:admin/open");
+        if (isAlreadyLegacy) {
+            return false; // this check was already made, so return false
+        }
+
+        log.info("Checking legacy permissions for {}", action);
+
+        action = action.replace(':', '/');
+        String opendistroAction = "cluster:admin/opendistro/" + action;
```

Comment:
I don't think we should assume that the name and legacy name of an action are overlapping. 

For AD, as an extension the transport action for get detector info is:

```
cluster:admin/opendistro/ad/detector/info
```

and will not be changed.

In this PR of the SDK (https://github.com/opensearch-project/opensearch-sdk-java/pulls) I introduced a concept of an `extensionShortName` (abbreviation) for an extension. For Anomaly Detection that could be `ad` like is done in this transport action name, but it could just as well be `anomaly` and then the REST API action name could be `anomaly:detector/info`. `detector/info` would be the `name` of the REST action corresponding to the detector info, but that name is entirely up to the extension developer and could change. Could we have a more flexible way of defining a legacy name?

Reply:
@DarshitChanpura FYI there are some transport action naming conventions in opensearch and you can find more details in this issue: https://github.com/opensearch-project/security/issues/2769#issuecomment-1544987612